### PR TITLE
feature: filter fifoWithBody

### PR DESCRIPTION
**Pull request by szuecs**
feature: filter fifoWithBody that works similar to fifo(), but release deferred until body streaming to client was finished

### CODE REVIEW

**Comment by szuecs** [comment]
@AlexanderYastrebov maybe better than https://github.com/zalando/skipper/pull/2675 wdyt?

File: filters/scheduler/fifo.go [hunk 1]
```diff
@@ -139,14 +164,26 @@ func (f *fifoFilter) Request(ctx filters.FilterContext) {
...
 func (f *fifoFilter) Response(ctx filters.FilterContext) {
-	pending, ok := ctx.StateBag()[fifoKey].([]func())
-	if !ok {
-		return
+	g := f.createResponse(ctx)
+	switch f.typ {
+	case fifo:
+		g()
+	case fifoWithBody:
+		ctx.StateBag()[filters.FifoWithBody] = g
```
Comment (AlexanderYastrebov) [inline filters/scheduler/fifo.go:172]: Its not clear how multiple fifoWithBody would chain and how they would interplay with regular fifo.
I think you'd need to have the same logic stored under different keys and then run chain of hooks in two places in proxy depending on the key at proxy.do and proxy.ServeHTTP:


```go
...
func (f *fifoFilter) createResponse(ctx filters.FilterContext, typ string) {
		pending, ok := ctx.StateBag()[typ].([]func())
		if !ok {
			return
		}
		last := len(pending) - 1
		if last < 0 {
			return
		}
		pending[last]()
		ctx.StateBag()[fifoKey] = pending[:last]
}

...
// existing in proxy.do
pendingFIFO, _ := stateBag[scheduler.FIFOKey].([]func())
for _, done := range pendingFIFO {
	done()
}
...
// in proxy.ServeHTTP
pendingFIFO, _ := stateBag[scheduler.FIFOWithBodyKey].([]func())
for _, done := range pendingFIFO {
	done()
}
```

File: filters/scheduler/fifo.go [hunk 2]
```diff
@@ -15,20 +15,44 @@ const (
 	fifoKey string = "fifo"
 )
 
+type fifoType int
+
+const (
+	fifo fifoType = iota + 1
+	fifoWithBody
+)
```
Comment (AlexanderYastrebov) [inline filters/scheduler/fifo.go:23]: You can use filter name as spec.typ and have withBody bool in filter similar to https://github.com/zalando/skipper/blob/2a202c47a7b0f80091f4492374b5c83c868ea9cb/filters/diag/histogram.go#L56-L57

Comment (szuecs) [reply filters/scheduler/fifo.go:23]: I like the enum style better than string.

File: filters/scheduler/fifo.go (hunk 1 above)
Comment (szuecs) [reply filters/scheduler/fifo.go:172]: Yes that's a good point. I also thought about that I need to do something like that.
What do you like more to support the copystream part? this one or the other?

Comment (AlexanderYastrebov) [reply filters/scheduler/fifo.go:172]: I think we can remove existing pendingFIFO from proxy.go see #2680

**Comment by szuecs** [comment]
close to create a PR to cw2023freeze

**Pull request closed without merging**